tests/commander.js

```diff
@@ -0,0 +1,180 @@
+            //{cmd: "choice before #form/group/select/item", type: "Choice", index: 2, indent: "    "},
+            //{cmd: "Choice after #form/group/select/item", type: "Choice", index: 3, indent: "    "},
+            {cmd: "choice", type: "Choice", index: 3, indent: "    ", select: "group/select"},
+            {cmd: "Choice in #form/group/select", type: "Choice", index: 3, indent: "    "},
+            {cmd: "Choice first in #form/group/select", type: "Choice", index: 2, indent: "    "},
+        ], function (args) {
+            it("should add a " + args.type + " question with '" + args.cmd + "'", function () {
+                util.paste([
+                    ["id", "type", "labelItext:en-default"],
+                    ["/group", "Group", "group"],
+                    ["/group/select", "Select", "select"],
+                    ["/group/select/item", "Choice", "item"],
+                    ["/text", "Text", "text"],
+                ]);
+                if (args.select) {
+                    util.clickQuestion(args.select);
+                }
+                var result = commander.doCommand(args.cmd, vellum);
+                assert.isOk(result, "question not added");
+                $("[name=property-nodeID]").val("new").change();
+                vellum.ensureCurrentMugIsSaved();
+                var items = [
+                    "group",
+                    "  select",
+                    "    item",
+                    "text",
+                ];
+                if (args.hasOwnProperty("index")) {
+                    items.splice(args.index, 0, (args.indent || "") + "new");
+                } else {
+                    items.push("new");
+                }
+                util.assertJSTreeState.apply(null, items);
+                assert.equal(result.__className, args.type);
+            });
+        });
+
+        _.each([
+            "#form/group",
+            "#form/group/select",
+            "#form/text",
+        ], function (path) {
+            it("should select " + path, function () {
+                util.paste([
+                    ["id", "type", "labelItext:en-default"],
+                    ["/group", "Group", "group"],
+                    ["/group/select", "Select", "select"],
+                    ["/group/select/item", "Choice", "item"],
+                    ["/text", "Text", "text"],
+                ]);
+                var result = commander.doCommand(path, vellum);
+                assert.isOk(result, "question not selected");
+                var mug = vellum.getCurrentlySelectedMug();
+                assert.equal(mug.hashtagPath, path);
+            });
+        });
+
+        // TODO decide if : is a good prefix for commands other than add or select
+        // TODO fix selection after delete
+        //  (two nodes selected in tree after delete non-adjacent to current)
+        // TODO delete multiple
+        //_.each([
+        //    ":delete",
```

Comment:
Haven't thought about this deeply, but one alternative would be to have different hotkeys for commands, instead of having a hotkey to open the commander and then you type a command. That would work with the current commands (add and jump), which both fit the pattern of opening the commander, typing a command, and then the commander closes and you need to re-open it. It wouldn't work so well if the commander becomes something more like a command-line interface that's used for multiple commands in a row.

Slack's command-K switcher is an elegant little in-between widget - it's definitely not a command-line interface, but it lets you do something more complex than a single hotkey would.